packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+#include "common/sort.h"
+#include "core/Context.h"
+#include "core/errors/packager.h"
+#include "core/errors/resolver.h"
+#include <iterator>
+
+namespace sorbet::packager {
+
+namespace {
+
+// For each __package.rb file, traverse the resolved tree and apply the visibility annotations to the symbols.
+class PropagateVisibility final {
+    const core::packages::PackageInfo &package;
+
+    bool definedByThisPackage(const core::GlobalState &gs, core::ClassOrModuleRef sym) {
+        auto pkg = gs.packageDB().getPackageNameForFile(sym.data(gs)->loc().file());
+        return this->package.mangledName() == pkg;
+    }
+
+    void recursiveExportSymbol(core::GlobalState &gs, bool firstSymbol, core::ClassOrModuleRef klass) {
+        // There is a loop between a class and its singleton, and this is the easiest way to detect it.
+        if (!firstSymbol && klass.data(gs)->flags.isExported) {
+            return;
+        }
+
+        // We only mark symbols from this package.
+        if (!this->definedByThisPackage(gs, klass)) {
+            return;
+        }
+
+        klass.data(gs)->flags.isExported = true;
+
+        for (const auto &child : klass.data(gs)->members()) {
+            if (child.second.isClassOrModule()) {
+                recursiveExportSymbol(gs, false, child.second.asClassOrModuleRef());
+            } else if (child.second.isFieldOrStaticField()) {
+                child.second.asFieldRef().data(gs)->flags.isExported = true;
+            }
+        }
+    }
+
+    void exportParentNamespace(core::GlobalState &gs, core::ClassOrModuleRef owner) {
+        while (owner.exists() && !owner.data(gs)->flags.isExported && this->definedByThisPackage(gs, owner)) {
+            owner.data(gs)->flags.isExported = true;
+            owner = owner.data(gs)->owner;
+        }
+    }
+
+    // Lookup the package name on the given root symbol, and mark the final symbol as exported.
+    void exportRoot(core::GlobalState &gs, core::ClassOrModuleRef sym) {
+        // Explicitly mark the symbol outside of the special package namespace as exported by this package, to allow
+        // other packages to refer to this namespace. Ultimately it would be great to not have this, but it helps with a
+        // migration to the visibility checker from the previous packages implementation.
+        for (auto name : this->package.fullName()) {
+            auto next = sym.data(gs)->findMember(gs, name);
+            if (!next.exists() || !next.isClassOrModule()) {
+                sym = core::Symbols::noClassOrModule();
+                break;
+            }
+
+            sym = next.asClassOrModuleRef();
+        }
+
+        if (sym.exists()) {
+            sym.data(gs)->flags.isExported = true;
+        }
+    }
+
+    // Mark both the package and its test namespace as exported. This will terminate the recursive exporting in
+    // `exportParentNamespace`, as it stops when it hits either the root or an exported symbol.
```

Comment:
```suggestion
    // Mark both `::A::B` and `::Test::A::B` as exported (given package `A::B`).
    // This will terminate the recursive exporting in `exportParentNamespace`, as it stops when it
    // hits either the root or an exported symbol.
```